src/lib/components/Search/index.js

```diff
@@ -0,0 +1,347 @@
+/**
+ * @fileoverview An Algolia search box.
+ */
+
+import {html} from "lit-element";
+import {BaseElement} from "../BaseElement";
+import {store} from "../../store";
+import {router} from "../../router";
+import algoliasearch from "algoliasearch/dist/algoliasearchLite";
+
+// Create an algolia client so we can get search results.
+// These keys are safe to be public.
+const applicationID = "2JPAZHQ6K7";
+const apiKey = "01ca870a3f1cad9984ed72419a12577c";
+const indexName = "webdev";
+const client = algoliasearch(applicationID, apiKey);
+const index = client.initIndex(indexName);
+
+/**
+ * An Algolia search box.
+ * @extends {BaseElement}
+ * @final
+ */
+class Search extends BaseElement {
+  static get properties() {
+    return {
+      // Manages the expanded/collapsed state of the UI.
+      expanded: {type: Boolean, reflect: true},
+      // An array of algolia results.
+      hits: {type: Object},
+      // Manages showing/hiding the search results popout.
+      showHits: {type: Boolean},
+      // Indicates which search result should be highlighted in the popout.
+      // Primarily used for keyboard behavior.
+      cursor: {type: Number},
+    };
+  }
+
+  constructor() {
+    super();
+    this.hits = [];
+    this.showHits = false;
+    this.cursor = -1;
+
+    // On smaller screens we don't do an animation so it's ok for us to fire off
+    // actions immediately. On larger screens we need to wait for the searchbox
+    // to fully expand/animate before we fire off actions.
+    // So we need to figure out our screen size and keep track of it if changes.
+    this.onResize = this.onResize.bind(this);
+    window.addEventListener("resize", this.onResize);
+    this.onResize();
+  }
+
+  render() {
+    return html`
+      <button
+        @click="${this.onOpenSearch}"
+        class="web-search__open-btn"
+        aria-label="Open search"
+      ></button>
+      <div
+        class="web-search__input-wrapper"
+        role="combobox"
+        aria-expanded="${this.expanded}"
+        aria-owns="web-search-popout__list"
+        aria-haspopup="listbox"
+      >
+        <input
+          class="web-search__input"
+          type="text"
+          role="search"
+          autocomplete="off"
+          aria-autocomplete="list"
+          aria-controls="web-search-popout__list"
+          aria-label="Search"
+          placeholder="Search"
+          @keyup="${this.onKeyUp}"
+          @focusin="${this.onFocusIn}"
+          @focusout="${this.onFocusOut}"
+        />
+      </div>
+      <button
+        @click="${this.onCloseSearch}"
+        class="web-search__close-btn"
+        aria-label="Close search"
+      ></button>
+      ${this.hitsTemplate}
+    `;
+  }
+
+  /* eslint-disable indent */
+  get hitsTemplate() {
+    return html`
```

Comment:
> Still getting my head around whether Lit would optimize this differently though (!?).

yeah that was my concern. I munged it all together to guarantee it would be "one template" to lit. I'll see if I can break it up without lit doing a weird thing.